### ISSUE TITLE
Introduce IAgent

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6702]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6702|
|Type|enhancement|
|Created|2013-08-26 14:29:16Z|
|Branch|https://github.com/twisted/twisted/tree/iagent-6702|

There are lots of Agent-alike classes and this was one of the specific goals of the Agent API.

To facilitate this, the Agent interface should be explicitly defined.

<details><summary>Searchable metadata</summary>

```
trac-id__6702 6702
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_iagent_6702 branches/iagent-6702
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1377527356000000 1377527356000000
changetime__1377897533000000 1377897533000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [39744]) Branching to 'iagent-6702'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

Not a code review but an @since marker is missing from the interface.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj set owner to @exarkun|
|-|-|

Hooray, an interface for Agent!

In addition to thijs's comment:

  1. There seems to be a minor grammatical error in the `IAgent` docstring: "or automatically response to authentication challenges".

  2. There are some new (annoying) twistedchecker errors, all involving the lack of `@return` fields in the `makeAgent` methods in the agent test cases.

  3. The documentation says that agents are intended to be composable but `ProxyAgent` does not accept an `agent` parameter, perhaps `ProxyAgent`'s docstring should mention that it is not composable and is intended to be at the bottom of an agent stack (the same as `Agent`.)

  4. Now that `Agent` implements `IAgent`, `Agent.request` ought to inherit the documentation from `IAgent.request`, do we need both pieces of documentation?

  5. `IAgent.request` names its URI parameter `url` while the implementation names it `uri`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

> Not a code review but an @since marker is missing from the interface

[#5309](https://github.com/twisted/twisted/issues/5309) :)

> There seems to be a minor grammatical error in the IAgent docstring: "or automatically response to authentication challenges".

Fixed in r39822.

> There are some new (annoying) twistedchecker errors, all involving the lack of @return fields in the makeAgent methods in the agent test cases.

Welp!  I fixed this in r39823 but the fix introduced _exactly_ the same number of 'definitely' spurious errors.  So, sort of an improvement, kind of.

> The documentation says that agents are intended to be composable but ProxyAgent does not accept an agent parameter, perhaps ProxyAgent's docstring should mention that it is not composable and is intended to be at the bottom of an agent stack (the same as Agent.)

Expanded on this idea in r39821.

> Now that Agent implements IAgent, Agent.request ought to inherit the documentation from IAgent.request, do we need both pieces of documentation?

Mmm, yes, sort of.  Made a change in r39826.

> IAgent.request names its URI parameter url while the implementation names it uri.

Hmm.  I have no idea which is correct.  I'll change the interface name since the implementation has been there for a long time already (r39828).

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_7"></a>|@jonathanj set owner to @exarkun|
|-|-|

One minor issue, your hypothetical browser-like code sample will not render as preformatted text because the block needs to begin with a double colon, http://epydoc.sourceforge.net/manual-epytext.html#literal-blocks.

Other than that, this looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [39833]) Merge iagent-6702

Author: exarkun
Reviewer: jonathanj
Fixes: [#6702](https://github.com/twisted/twisted/issues/6702)

Introduce `twisted.web.iweb.IAgent`, an explicit interface definition
for the implicit interface already implemented by several "Agent"
classes in `twisted.web.client`.
